gordon_janitor_gcp/gce_instance_authority.py

```diff
@@ -0,0 +1,145 @@
+            messages.
+    """
+    def __init__(self, config, rrset_channel=None, **kwargs):
+        self.config = config
+        self.keyfile = self.config['keyfile']
+        self.rrset_channel = rrset_channel
+
+    async def get_all_projects(self):
+        """Get all active GCE projects.
+
+        Returns:
+            set: active GCE project IDs.
+        """
+        auth_client = auth.GoogleAuthClient(
+            self.keyfile,
+            scopes=['cloud-platform.read-only'],
+            session=self.session)
+        crm_client = gce_clients.CRMClient(auth_client, self.session)
+
+        active_projects = await crm_client.list_all_active_projects()
+        return set(p.get('projectId') for p in active_projects)
+
+    async def get_processed_projects(self):
+        """Get list of GCE project IDs with blacklisted ones filtered out.
+
+        Returns:
+            list: project IDs.
+        """
+        projects = await self.get_all_projects()
+        project_blacklist = set(self.config['project_blacklist'])
+        return sorted(projects - project_blacklist)
+
+    async def get_instance_filter(self):
+        """Get filter string used by instances.aggregatedList endpoint.
+
+        For filter string details, see api endpoint documentation:
+        https://cloud.google.com/compute/docs/reference/latest/instances/aggregatedList
+
+        Returns:
+            str: filter suitable for the 'filter' endpoint parameter.
+        """
+        return ''
+
+    async def get_instances_for_projects(self, projects):
+        """Get list of instance data for each project.
+
+        Args:
+            projects (list): list of project names.
+        Yields:
+            list: list of dicts that contain instance information.
+        """
+        tag_blacklist = self.config.get('tag_blacklist', [])
+        metadata_blacklist = self.config.get('metadata_blacklist', {})
+
+        auth_client = auth.GoogleAuthClient(
+            self.keyfile, scopes=['cloud-platform'], session=self.session)
+
+        gcp_client = gce_clients.GCPClient(
+            auth_client,
+            self.session,
+            tag_blacklist=tag_blacklist,
+            metadata_blacklist=metadata_blacklist)
+
+        instance_filter = await self.get_instance_filter()
+
+        for project in projects:
+            yield await gcp_client.list_instances(
+                project, instance_filter=instance_filter)
+
+    async def get_msg_generator(self):
+        """Get message generator to translate instance data into channel msgs.
+
+        Returns:
+            obj: obj that creates messages for the channel.
+        """
+        return msg_generators.ARecordMsgGenerator(self.config['zone'])
+
+    async def start(self):
+        """Gather instance data batches and send them to the
+        :py:obj:`self.rrset_channel`.
+        """
+        async with aiohttp.ClientSession() as session:
+            self.session = session
```

Comment:
I'm interested in why this is set with an async context manager, which I also saw in the all the examples in the docs, but I couldn't find the reasoning for it. I understand session requests need to be async, but not necessarily why establishing the session needs to be

Reply:
I had a similar/related question too - but it's also a question of the bigger picture. I know that using a context manager with `aiohttp.ClientSession` automatically closes the session once you're done with the context/outside the scope. But one could also directly close the client session upon cleanup of the client (i.e. emit a `None` message after all messages sent, session._close).

I initially was curious why creating a new client session for this, since an auth client will create one, and the http client will automatically reuse the auth client one. I then realized that because of all the iteration going on here (`async for project_instances in self.get_instances_for_projects`, then `for project in projects`, then iterating over pages, then all instances of the complete response - loooots of iteration), and sort of forcing you to create a new instance of the auth client, and GCE client, and the CRM client. So then the questions has morphed into: can't we create one instance of the three clients, and reuse them all? I don't think these new instances of clients are different every time (unless I'm completely missing on something). So no need for creating an explicit session, no need for a new object every iteration. Could the three just be created upon instantiation of the Authority class? Similar to how the reconciler currently does it, and the proposed pubsub client does.

Reply:
I think the reason why this is using an `async with` statement is because closing the session is an async operation, so an ordinary context manager would not be able to call the [async close](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/client.py#L696) method.

Why use a context manager at all here? I think it's a choice between that and a `try...except...finally` block because it'd be nice if the authority closed the session when it finishes and when an uncaught exception occurs. I couldn't think of anything the try-except block could catch, except maybe a generic `except Exception`, so I went with `with`. 

> can't we create one instance of the three clients, and reuse them all?

Can you point out where a new object is created on very iteration? In my understanding, there are only 4 client-like objects: 2 auth clients, 1 CRMClient, and 1 GCEClient. Each client is reused after instantiation, and each of them uses the same `aiothttp.ClientSession`.
Also, can one auth client be shared between clients that use different scopes and service accounts?

Also, this made me notice that we never await `session.close()` calls in the reconciler. Not entirely sure what this means, but could be something like leaving sockets open, so we should patch that up later.

Reply:
Ah you're right re: not multiple clients - that's what I get for reviewing while jet lagged. It was just confusing with the generator. 

Good call on the `await` missing. I don't think I'll have time to make a quick PR to fix so if you want to, by all means :)

I do think the argument for init'ing all the clients when the authority plugin is init'ed is probably still valid, though, right?

Gah I should get some sleep.

Reply:
Thanks for the explanation to my question! I had also forgotten that the way we plan to have multiple janitors in one process. I was thinking the program would end immediately after start() was called, and since we've toyed with the idea of running this on a 5 min cron, I figured if any resources were claimed, they would be freed by then, so I didn't really see the point of calling close (I may have some other misconceptions there, too, but it seems like close doesn't do *that* much https://aiohttp.readthedocs.io/en/stable/client_advanced.html#graceful-shutdown). It was also my understanding that close() was not an async method, since you can call it synchronously and the docs don't mention one way or the other. However, now that I've looked at the source code (https://github.com/aio-libs/aiohttp/blob/master/aiohttp/client.py#L696), I see that it is defined as async; it just doesn't raise any exceptions if you call it synchronously.

Reply:
> I do think the argument for init'ing all the clients when the authority plugin is init'ed is probably still valid, though, right?

@econchick In this instance, initing the clients in specific methods makes it easy to override these methods and put some internal-only logic into them - if I moved the client inits to the plugin's init, then adding internal logic would require more changes (eg. would have to override __init__(), the method in question, and possibly start()). 

This assumes that the internal-facing plugin package will depend on this package so that it can import and override some things. Is there some other way to handle the public/internal split that I'm missing? We can talk about it on slack if that makes it easier.

Reply:
@descent3me I think that's a good point. If I remember correctly, the aim is to kick off the janitor and let it run for a few minutes to do its job, in which case you're right that we could leave this stuff hanging and trust the OS to clean up any sockets. I'll think a bit more what this means if we have >1 ClientSessions and see which way to go.

I dug in deeper into the `close` stuff, and [it never does anything asynchronous](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/connector.py#L314) later as far as I can tell, so there's no need for the async context manager, unless the library author is planning something for the future. That said, if an async method gets called but not awaited, it never gets scheduled for execution and the interpreter issues a warning, so all those calls to `session.close()` are noops :):

```pycon
In [6]: import asyncio

In [7]: async def work():
   ...:     print('working')
   ...:     await asyncio.sleep(1)
   ...:     
   ...:     

In [8]: async def main():
   ...:     work()
   ...:     await work()
   ...:     

In [9]: loop = asyncio.get_event_loop()

In [10]: loop.run_until_complete(main())
/home/matto/.virtualenvs/sandbox/bin/ipython:2: RuntimeWarning: coroutine 'work' was never awaited
  
working
```